File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode[]} lists
     * @return {ListNode}
     */
    mergeKLists(lists) {
        if(lists.length == 0 || lists == null){
            return null;
        }
        while(lists.length > 1){
            const mergedLists = [];
            for(let i =0; i < lists.length; i+=2){
                const list1 = lists[i];
                const list2 = i + 1 < lists.length ? lists[i + 1] : null;
                mergedLists.push(this.mergeLists(list1, list2));
            }
            lists = mergedLists;
        }
        return lists[0];
    }

    mergeLists(l1, l2){
        let mergedList = new ListNode(null);
        let head = mergedList;
        if(l1 == null && l2 == null){
            return mergedList.next;
        }
        
        while(l1 !== null && l2 !== null){
            if(l1.val <= l2.val){
                mergedList.next = l1;
                mergedList = mergedList.next;
                l1 = l1.next;
            } else {
                mergedList.next = l2;
                mergedList = mergedList.next;
                l2 = l2.next;
            }
        }
        if(l1 != null){
            mergedList.next = l1;
        } else if(l2 != null) {
           mergedList.next = l2;
        }
        return head.next;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next        
        second = slow.next
        slow.next = None
        prev = slow.next
        while second:
            temp = second.next #save the link
            second.next = prev #set
            prev = second
            second = temp
        first = head
        second = prev
        while first and second:
            temp1 = first.next
            temp2 = second.next
            first.next = second
            second.next = temp1
            first, second = temp1, temp2

